File: src/services/inference/recheck_stats.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

let per_file_time_guess = 0.003

let per_merge_file_time_guess = 0.001

let per_check_file_time_guess = 0.005

let per_file_time_key = "per_file_time"

let per_merge_file_time_key = "per_merge_file_time"

let per_check_file_time_key = "per_check_file_time"

let estimates_key = "estimates"

let estimated_time_to_recheck_key = "estimated_time_to_recheck"

let estimated_time_to_restart_key = "estimated_time_to_restart"

let estimated_time_to_init_key = "estimated_time_to_init"

let estimated_time_per_file_key = "estimated_time_per_file"

let estimated_files_to_recheck_key = "estimated_files_to_recheck"

let estimated_files_to_init_key = "estimated_files_to_init"

type estimates = {
  estimated_time_to_recheck: float;
  estimated_time_to_restart: float;
  estimated_time_to_init: float;
  estimated_time_per_file: float;
  estimated_files_to_recheck: int;
  estimated_files_to_init: int;
}

type averages = {
  init_time: float;
  per_file_time: float;
  per_merge_file_time: float;
  per_check_file_time: float;
  parsed_count: int;
}

let averages = ref None

(* window should be a positive integer *)
let moving_average ~window ~avg ~sample ~sample_count =
  let window = float_of_int window in
  let sample_count = float_of_int sample_count in
  if sample_count >= window then
    sample
  else
    ((avg *. (window -. sample_count)) +. (sample *. sample_count)) /. window

let get_file ~options =
  let root = Options.root options in
  let tmp_dir = Options.temp_dir options in
  let flowconfig_name = Options.flowconfig_name options in
  Server_files_js.recheck_stats_file ~flowconfig_name ~tmp_dir root

let load_stats ~options =
  Lwt_result.(
    let file = get_file ~options in
    let%lwt result =
      (try%lwt
         Lwt_result.ok
           (Lwt_io.open_file
              ~flags:[Unix.O_RDONLY; Unix.O_NONBLOCK]
              ~mode:Lwt_io.Input
              ~perm:0o666
              file
           )
       with
      | Unix.Unix_error (Unix.ENOENT, _, _) -> Lwt_result.fail "File doesn't exist")
      >>= fun ic ->
      let%lwt contents = Lwt_io.read ic in
      let%lwt () = Lwt_io.close ic in
      Lwt_result.lift
        (try
           let json = Some (Hh_json.json_of_string contents) in
           let per_file_time =
             match Hh_json_helpers.Jget.float_opt json per_file_time_key with
             | Some v -> v
             | None -> per_file_time_guess
           in
           let per_merge_file_time =
             match Hh_json_helpers.Jget.float_opt json per_merge_file_time_key with
             | Some v -> v
             | None -> per_merge_file_time_guess
           in
           let per_check_file_time =
             match Hh_json_helpers.Jget.float_opt json per_check_file_time_key with
             | Some v -> v
             | None -> per_check_file_time_guess
           in
           Result.Ok (per_file_time, per_merge_file_time, per_check_file_time)
         with
        | Hh_json.Syntax_error str ->
          Result.Error (Printf.sprintf "Failed to parse as JSON contents. %S: %S" str contents)
        | Hh_json_helpers.Jget.Parse key ->
          Result.Error (Printf.sprintf "Failed to find key %S in estimates object. %S" key contents))
    in
    match result with
    | Result.Ok (per_file_time, per_merge_file_time, per_check_file_time) ->
      Lwt.return (per_file_time, per_merge_file_time, per_check_file_time)
    | Result.Error reason ->
      Hh_logger.info "Failed to load recheck stats from %S. Reason: %S" file reason;
      Lwt.return (per_file_time_guess, per_merge_file_time_guess, per_check_file_time_guess)
  )

let save_averages ~options ?estimates new_averages =
  averages := Some new_averages;

  let estimates =
    Base.Option.value_map
      estimates
      ~default:[]
      ~f:(fun
           {
             estimated_time_to_recheck;
             estimated_time_to_restart;
             estimated_time_to_init;
             estimated_time_per_file;
             estimated_files_to_recheck;
             estimated_files_to_init;
           }
         ->
        [
          ( estimates_key,
            Hh_json.JSON_Object
              [
                ( estimated_time_to_recheck_key,
                  Hh_json.JSON_Number (Dtoa.ecma_string_of_float estimated_time_to_recheck)
                );
                ( estimated_time_to_restart_key,
                  Hh_json.JSON_Number (Dtoa.ecma_string_of_float estimated_time_to_restart)
                );
                ( estimated_time_to_init_key,
                  Hh_json.JSON_Number (Dtoa.ecma_string_of_float estimated_time_to_init)
                );
                ( estimated_time_per_file_key,
                  Hh_json.JSON_Number (Dtoa.ecma_string_of_float estimated_time_per_file)
                );
                ( estimated_files_to_recheck_key,
                  Hh_json.JSON_Number (string_of_int estimated_files_to_recheck)
                );
                ( estimated_files_to_init_key,
                  Hh_json.JSON_Number (string_of_int estimated_files_to_init)
                );
              ]
          );
        ]
    )
  in
  let json_str =
    Hh_json.(
      json_to_string
      @@ JSON_Object
           ([
              (per_file_time_key, JSON_Number (Dtoa.ecma_string_of_float new_averages.per_file_time));
              ( per_merge_file_time_key,
                JSON_Number (Dtoa.ecma_string_of_float new_averages.per_merge_file_time)
              );
              ( per_check_file_time_key,
                JSON_Number (Dtoa.ecma_string_of_float new_averages.per_check_file_time)
              );
            ]
           @ estimates
           )
    )
  in
  let file = get_file ~options in
  Lwt_result.(
    let%lwt result =
      (try%lwt
         Lwt_result.ok
         @@ Lwt_io.open_file
              ~flags:[Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC]
              ~mode:Lwt_io.Output
              ~perm:0o666
              file
       with
      | Unix.Unix_error (Unix.ENOENT, _, _) -> Lwt_result.fail "File doesn't exist"
      | exn ->
        let exn = Exception.wrap exn in
        Lwt_result.fail (Printf.sprintf "Failed to open file\n%s" (Exception.to_string exn)))
      >>= fun oc ->
      try%lwt
        let%lwt () = Lwt_io.write oc json_str in
        Lwt_result.ok @@ Lwt_io.close oc
      with
      | exn ->
        let exn = Exception.wrap exn in
        Lwt_result.fail (Printf.sprintf "Failed to write file\n%s" (Exception.to_string exn))
    in
    begin
      match result with
      | Result.Ok () -> ()
      | Result.Error msg -> Hh_logger.error "Failed to save per_file_time to %S. %s" file msg
    end;

    Lwt.return_unit
  )

let init ~options ~init_time ~parsed_count =
  let%lwt (per_file_time, per_merge_file_time, per_check_file_time) = load_stats ~options in
  averages :=
    Some { init_time; per_file_time; per_merge_file_time; per_check_file_time; parsed_count };
  Lwt.return_unit

let with_averages f =
  match !averages with
  | None -> failwith "Recheck_stats needs to be initialized before it can be used"
  | Some averages -> f averages

let record_recheck_time ~options ~merge_time ~check_time ~merged_files ~checked_files =
  with_averages
  @@ fun { init_time; per_file_time; per_merge_file_time; per_check_file_time; parsed_count } ->
  (* What should we do for tiny repositories? Let's make the window at least 15 samples big *)
  let window = max parsed_count 15 in
  let total_files = merged_files + checked_files in
  let per_file_time =
    if total_files > 0 then
      moving_average
        ~window
        ~avg:per_file_time
        ~sample:((merge_time +. check_time) /. float_of_int total_files)
        ~sample_count:total_files
    else
      per_file_time
  in
  let per_merge_file_time =
    if merged_files > 0 then
      moving_average
        ~window
        ~avg:per_merge_file_time
        ~sample:(merge_time /. float_of_int merged_files)
        ~sample_count:merged_files
    else
      per_merge_file_time
  in
  let per_check_file_time =
    if checked_files > 0 then
      moving_average
        ~window
        ~avg:per_check_file_time
        ~sample:(check_time /. float_of_int checked_files)
        ~sample_count:checked_files
    else
      per_check_file_time
  in
  save_averages
    ~options
    { init_time; per_file_time; per_merge_file_time; per_check_file_time; parsed_count }

let record_last_estimates ~options ~estimates =
  with_averages @@ fun averages -> save_averages ~options ~estimates averages

let get_init_time () = with_averages @@ fun { init_time; _ } -> init_time

let get_per_merge_file_time () =
  with_averages @@ fun { per_merge_file_time; _ } -> per_merge_file_time

let get_per_check_file_time () =
  with_averages @@ fun { per_check_file_time; _ } -> per_check_file_time

File: rust_port/crates/flow_services_inference/src/recheck_stats.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;

use flow_common::options::Options;
use flow_server_files::server_files_js;

const PER_FILE_TIME_GUESS: f64 = 0.003;

const PER_MERGE_FILE_TIME_GUESS: f64 = 0.001;

const PER_CHECK_FILE_TIME_GUESS: f64 = 0.005;

const PER_FILE_TIME_KEY: &str = "per_file_time";

const PER_MERGE_FILE_TIME_KEY: &str = "per_merge_file_time";

const PER_CHECK_FILE_TIME_KEY: &str = "per_check_file_time";

const ESTIMATES_KEY: &str = "estimates";

const ESTIMATED_TIME_TO_RECHECK_KEY: &str = "estimated_time_to_recheck";

const ESTIMATED_TIME_TO_RESTART_KEY: &str = "estimated_time_to_restart";

const ESTIMATED_TIME_TO_INIT_KEY: &str = "estimated_time_to_init";

const ESTIMATED_TIME_PER_FILE_KEY: &str = "estimated_time_per_file";

const ESTIMATED_FILES_TO_RECHECK_KEY: &str = "estimated_files_to_recheck";

const ESTIMATED_FILES_TO_INIT_KEY: &str = "estimated_files_to_init";

pub struct Estimates {
    pub estimated_time_to_recheck: f64,
    pub estimated_time_to_restart: f64,
    pub estimated_time_to_init: f64,
    pub estimated_time_per_file: f64,
    pub estimated_files_to_recheck: i64,
    pub estimated_files_to_init: i64,
}

struct Averages {
    init_time: f64,
    per_file_time: f64,
    // per_merge_file_time: float;
    per_merge_file_time: f64,
    // per_check_file_time: float;
    per_check_file_time: f64,
    parsed_count: i64,
}

static AVERAGES: Mutex<Option<Averages>> = Mutex::new(None);

// window should be a positive integer
fn moving_average(window: i64, avg: f64, sample: f64, sample_count: i64) -> f64 {
    let window = window as f64;
    let sample_count = sample_count as f64;
    if sample_count >= window {
        sample
    } else {
        ((avg * (window - sample_count)) + (sample * sample_count)) / window
    }
}

fn get_file(options: &Options) -> PathBuf {
    let root = &options.root;
    let tmp_dir = &options.temp_dir;
    let flowconfig_name = &options.flowconfig_name;
    PathBuf::from(server_files_js::recheck_stats_file(
        flowconfig_name,
        tmp_dir,
        root,
    ))
}

fn load_stats(options: &Options) -> (f64, f64, f64) {
    let file = get_file(options);
    let result: Result<(f64, f64, f64), String> = (|| {
        let contents = fs::read_to_string(&file).map_err(|e| {
            if e.kind() == std::io::ErrorKind::NotFound {
                "File doesn't exist".to_string()
            } else {
                format!("Failed to open file: {}", e)
            }
        })?;
        let json: serde_json::Value = serde_json::from_str(&contents).map_err(|e| {
            format!(
                "Failed to parse as JSON contents. {:?}: {:?}",
                e.to_string(),
                contents
            )
        })?;
        let per_file_time = json
            .get(PER_FILE_TIME_KEY)
            .and_then(|v| v.as_f64())
            .unwrap_or(PER_FILE_TIME_GUESS);
        let per_merge_file_time = json
            .get(PER_MERGE_FILE_TIME_KEY)
            .and_then(|v| v.as_f64())
            .unwrap_or(PER_MERGE_FILE_TIME_GUESS);
        let per_check_file_time = json
            .get(PER_CHECK_FILE_TIME_KEY)
            .and_then(|v| v.as_f64())
            .unwrap_or(PER_CHECK_FILE_TIME_GUESS);
        Ok((per_file_time, per_merge_file_time, per_check_file_time))
    })();
    match result {
        Ok((per_file_time, per_merge_file_time, per_check_file_time)) => {
            (per_file_time, per_merge_file_time, per_check_file_time)
        }
        Err(reason) => {
            flow_hh_logger::info!(
                "Failed to load recheck stats from {:?}. Reason: {:?}",
                file.display(),
                reason
            );
            (
                PER_FILE_TIME_GUESS,
                PER_MERGE_FILE_TIME_GUESS,
                PER_CHECK_FILE_TIME_GUESS,
            )
        }
    }
}

fn save_averages(options: &Options, estimates: Option<&Estimates>, new_averages: Averages) {
    let per_file_time = new_averages.per_file_time;
    let per_merge_file_time = new_averages.per_merge_file_time;
    let per_check_file_time = new_averages.per_check_file_time;

    {
        let mut guard = AVERAGES.lock().unwrap();
        *guard = Some(new_averages);
    }

    let mut json_obj = serde_json::Map::new();
    json_obj.insert(
        PER_FILE_TIME_KEY.to_string(),
        serde_json::Value::from(per_file_time),
    );
    json_obj.insert(
        PER_MERGE_FILE_TIME_KEY.to_string(),
        serde_json::Value::from(per_merge_file_time),
    );
    json_obj.insert(
        PER_CHECK_FILE_TIME_KEY.to_string(),
        serde_json::Value::from(per_check_file_time),
    );

    if let Some(Estimates {
        estimated_time_to_recheck,
        estimated_time_to_restart,
        estimated_time_to_init,
        estimated_time_per_file,
        estimated_files_to_recheck,
        estimated_files_to_init,
    }) = estimates
    {
        let mut estimates_obj = serde_json::Map::new();
        estimates_obj.insert(
            ESTIMATED_TIME_TO_RECHECK_KEY.to_string(),
            serde_json::Value::from(*estimated_time_to_recheck),
        );
        estimates_obj.insert(
            ESTIMATED_TIME_TO_RESTART_KEY.to_string(),
            serde_json::Value::from(*estimated_time_to_restart),
        );
        estimates_obj.insert(
            ESTIMATED_TIME_TO_INIT_KEY.to_string(),
            serde_json::Value::from(*estimated_time_to_init),
        );
        estimates_obj.insert(
            ESTIMATED_TIME_PER_FILE_KEY.to_string(),
            serde_json::Value::from(*estimated_time_per_file),
        );
        estimates_obj.insert(
            ESTIMATED_FILES_TO_RECHECK_KEY.to_string(),
            serde_json::Value::from(*estimated_files_to_recheck),
        );
        estimates_obj.insert(
            ESTIMATED_FILES_TO_INIT_KEY.to_string(),
            serde_json::Value::from(*estimated_files_to_init),
        );
        json_obj.insert(
            ESTIMATES_KEY.to_string(),
            serde_json::Value::Object(estimates_obj),
        );
    }

    let json_str = serde_json::to_string(&serde_json::Value::Object(json_obj))
        .expect("Failed to serialize JSON");

    let file = get_file(options);

    let result: Result<(), String> = fs::write(&file, &json_str).map_err(|e| {
        if e.kind() == std::io::ErrorKind::NotFound {
            "File doesn't exist".to_string()
        } else {
            format!("Failed to open file\n{}", e)
        }
    });

    match result {
        Ok(()) => {}
        Err(msg) => {
            flow_hh_logger::error!(
                "Failed to save per_file_time to {:?}. {}",
                file.display(),
                msg
            );
        }
    }
}

pub fn init(options: &Options, init_time: f64, parsed_count: i64) {
    let (per_file_time, per_merge_file_time, per_check_file_time) = load_stats(options);
    let mut guard = AVERAGES.lock().unwrap();
    *guard = Some(Averages {
        init_time,
        per_file_time,
        per_merge_file_time,
        per_check_file_time,
        parsed_count,
    });
}

fn with_averages<T>(f: impl FnOnce(&Averages) -> T) -> T {
    let guard = AVERAGES.lock().unwrap();
    match &*guard {
        None => panic!("Recheck_stats needs to be initialized before it can be used"),
        Some(averages) => f(averages),
    }
}

pub fn record_recheck_time(
    options: &Options,
    merge_time: f64,
    check_time: f64,
    merged_files: i64,
    checked_files: i64,
) {
    // with_averages
    let (init_time, per_file_time, per_merge_file_time, per_check_file_time, parsed_count) =
        with_averages(|averages| {
            let Averages {
                init_time,
                per_file_time,
                per_merge_file_time,
                per_check_file_time,
                parsed_count,
            } = *averages;
            (
                init_time,
                per_file_time,
                per_merge_file_time,
                per_check_file_time,
                parsed_count,
            )
        });
    // What should we do for tiny repositories? Let's make the window at least 15 samples big
    let window = parsed_count.max(15);
    let total_files = merged_files + checked_files;
    let per_file_time = if total_files > 0 {
        moving_average(
            window,
            per_file_time,
            (merge_time + check_time) / total_files as f64,
            total_files,
        )
    } else {
        per_file_time
    };
    let per_merge_file_time = if merged_files > 0 {
        moving_average(
            window,
            per_merge_file_time,
            merge_time / merged_files as f64,
            merged_files,
        )
    } else {
        per_merge_file_time
    };
    let per_check_file_time = if checked_files > 0 {
        moving_average(
            window,
            per_check_file_time,
            check_time / checked_files as f64,
            checked_files,
        )
    } else {
        per_check_file_time
    };
    save_averages(
        options,
        None,
        Averages {
            init_time,
            per_file_time,
            per_merge_file_time,
            per_check_file_time,
            parsed_count,
        },
    );
}

pub fn record_last_estimates(options: &Options, estimates: &Estimates) {
    let (init_time, per_file_time, per_merge_file_time, per_check_file_time, parsed_count) =
        with_averages(|averages| {
            let Averages {
                init_time,
                per_file_time,
                per_merge_file_time,
                per_check_file_time,
                parsed_count,
            } = *averages;
            (
                init_time,
                per_file_time,
                per_merge_file_time,
                per_check_file_time,
                parsed_count,
            )
        });
    save_averages(
        options,
        Some(estimates),
        Averages {
            init_time,
            per_file_time,
            per_merge_file_time,
            per_check_file_time,
            parsed_count,
        },
    );
}

pub fn get_init_time() -> f64 {
    with_averages(|Averages { init_time, .. }| *init_time)
}

pub fn get_per_merge_file_time() -> f64 {
    with_averages(
        |Averages {
             per_merge_file_time,
             ..
         }| *per_merge_file_time,
    )
}

pub fn get_per_check_file_time() -> f64 {
    with_averages(
        |Averages {
             per_check_file_time,
             ..
         }| *per_check_file_time,
    )
}
